[RFC] Add new test: fi_msg_sockets

patrickmacarthur opened the pull request:
This test verifies that using fi_setname() on a passive endpoint and transforming the passive endpoint into an active one on the client side works correctly.  This functionality is required by UNH EXS and rsockets.

I am open to suggestions on how to better integrate this into fabtests, but having had multiple provider maintainers use this test, I would like to get something to test this particular functionality into fabtests proper.

This has been tested against ofiwg/libfabric#1160 and contains the fix that @a-ilango requested in that PR.

patrickmacarthur commented:
This PR needs to be updated after the last update to PR ofiwg/libfabric#1160.  It enforces that the address format passed to `fi_setname()` must match `fi->addr_format` returned by `fi_getinfo()`.  Thus, we need to specify the exact address format we want in hints.

I will make this update to the PR tomorrow.

patrickmacarthur commented:
Just as an update: I added commits to set the `addr_format` in hints properly and to close the passive endpoint after creating an active endpoint from it.

shefty commented:
This looks fine to me.  Want me to merge?

goodell commented:
@bturrubiates and I are trying this test today with the usnic provider and seeing some some strange behavior.  We're still trying to figure out if the test is doing the right thing or our provider has some bug.

shefty commented on simple/msg_setname.c:
```diff
@@ -0,0 +1,647 @@
+		goto free_pep;
+	} else if (bound_addr->sa_family != AF_INET) {
+		FT_ERR("got unexpected address family %d\n",
+				bound_addr->sa_family);
```
There is a check above that sets hints->addr_format to INET, IB, or INET6, but this only handles the case for INET.  Something is off here.

patrickmacarthur replied:
_Trying this reply again since the Android e-mail client apparently strips out all headers on reply :(_

That was a debug check that was accidentally left in.  Will fix.

shefty commented:
Taking a second look, I do see one issue with the test.  What was the problem that you were seeing with it?

goodell commented:
`./fi_msg_setname -f usnic` doesn't seem to work on our systems, but `./fi_msg_setname -f usnic -s $FIRST_VALID_ADDRESS` does.  Also, `./fi_msg_setname -f usnic -s $SECOND_VALID_ADDRESS` does not.

bturrubiates commented on simple/msg_setname.c:
```diff
@@ -0,0 +1,647 @@
+	hints->addr_format	= FI_SOCKADDR;
+
+	/* Fabric and connection setup */
+	if (opts.src_addr || opts.src_port) {
```
What is the purpose of this conditional? Should this only be run for the server? Both server and client will have `opts.src_port` set since they both use `INIT_OPTS` for initialization.

shefty replied:
If this test isn't true, I don't think the test will run correctly.  setup_handle() needs to be called on both sides.  So, I think this check should be: if it's not true, abort the test.

bturrubiates replied:
It seems strange to me that the client would call getinfo in `setup_handle` then again in `client_connect`.

shefty replied:
Where is setup_handle called in client_connect?

goodell replied:
I think Ben was saying that it was strange to him for the client to call **`fi_getinfo`** twice, not that `client_connect` is called twice.

shefty replied:
The test is trying to mimic the use case you would get implement sockets over libfabric.  So the app allocates a passive endpoint, binds an address to it, then 'migrates' the endpoint from passive to active.  The name of the test isn't conveying this, so a better name may help here.

patrickmacarthur commented on simple/msg_setname.c:
```diff
@@ -0,0 +1,647 @@
+	}
+
+	/* Get fabric info */
+	ret = fi_getinfo(FT_FIVERSION, NULL, NULL, 0, hints, &fi);
```
Before I went on vacation, I had noticed that this call should probably have `FI_SOURCE` set in flags, since we are creating a passive endpoint from it.  However, when I made this fix, it caused the sockets provider to fail here, and I ran out of time to investigate it.  Maybe this is the cause of the usNIC provider not working with this test?

I will look into the root cause of this issue today.

patrickmacarthur replied:
@shefty Is `fi_getinfo(FT_FIVERSION, NULL, NULL, FI_SOURCE, hints, &fi);` with `src_addr` and `dest_addr` in hints NULL'd out a sane/correct thing to do in this case?

shefty replied:
> @shefty <https://github.com/shefty>  Is fi_getinfo(FT_FIVERSION, NULL,
> NULL, FI_SOURCE, hints, &fi); with src_addr and dest_addr in hints NULL'd
> out a sane/correct thing to do in this case?

Adding the FI_SOURCE flag doesn't help any, since you haven't specified a source address anywhere. So, I would vote no, it doesn't really make sense.

patrickmacarthur replied:
OK, thanks for the clarification.  I will leave this as-is.

bturrubiates commented:
@patrickmacarthur @shefty @goodell 
To clarify on the behavior we're seeing: 

The usNIC provider maps fabrics to devices. Since this test calls `fi_getinfo` without any arguments related to addressing info it has an implicitly assigned address because it can only use whatever device was first returned in the list of devices. This means that at least for the usNIC provider we can't properly handle any device other than the first working device, even if we provide a `-s` parameter. For it to work properly for us you need to provide an address in the `fi_getinfo` calls so that the devices we map the fabric to can properly resolve the addresses requested. 

If I change it to something more like this: 

```
        /* Get fabric info */
-       ret = fi_getinfo(FT_FIVERSION, NULL, NULL, 0, hints, &fi);
+       if (opts.dst_addr)
+               ret = fi_getinfo(FT_FIVERSION, opts.dst_addr, NULL, 0, hints, &fi);
+       else
+               ret = fi_getinfo(FT_FIVERSION, opts.src_addr, NULL, FI_SOURCE, hints, &fi);
+
```

This modification allows us to properly return a device that can resolve the addresses, but it might defeat the purpose of the test.

Does that make sense?

shefty commented:
This makes sense.  The problem seems like an issue with the test.  It can't assume that the address passed to setname is usable with the opened domain.

patrickmacarthur commented:
@shefty So, what I should do here then is (1) open all possible fabric handles (that other than source address match the user's requested parameters) at the start, then (2) for each fabric handle, try opening a passive endpoint and then calling `fi_setname()` until one succeeds?

If not, I'm seriously misunderstanding the use case of `fi_setname()`.

As an aside, UNH EXS currently does not use `fi_setname()` at all; it instead passes the source address directly to `fi_getinfo()` at `exs_bind()` time.  So the only (intentional) difference between this test and what UNH EXS does right now is that EXS doesn't call `fi_setname()`.

shefty commented:
An endpoint cannot be assigned an arbitrary address.  It has to be one that can be associated with a specific domain.

You can use fi_setname to assign a port to an endpoint, if it were allocated without one.

What we ultimately may need are APIs to discover which addresses are associated with what domains.

For a test app, I guess we could call fi_getinfo() with a specific address.  Remove that address from the resulting fi_info structure.  Allocate an endpoint.  Call setname using the address that was removed. ?

goodell commented:
> For a test app, I guess we could call fi_getinfo() with a specific address.  Remove that address from the resulting fi_info structure.  Allocate an endpoint.  Call setname using the address that was removed. ?

Yes, this makes sense to me.

patrickmacarthur commented:
Added a couple of commits which implement @shefty's suggested flow, and renamed the test to better indicate its purpose.  Tested on verbs and sockets providers (I don't have access to usNIC hardware).

Let me know whether or not the updated test works/makes sense.  I would like to squash the commits and fix up some whitespace issues before this gets merged.

goodell commented on simple/msg_sockets.c:
```diff
@@ -0,0 +1,664 @@
+		FT_PRINTERR("fi_getinfo", ret);
+		goto free_ai;
+	}
+	fi->src_addr = NULL;
```
nit: does this result in a minor memory leak from the memory allocated as part of `fi_allocinfo`?

patrickmacarthur replied:
Fixed.  I believe I can free it unconditionally since it will either be NULL (in which case `free` is a no-op) or a malloc'd buffer (since `fi_freeinfo` also unconditionally frees it).

goodell commented:
@bturrubiates would please you re-test @patrickmacarthur's latest update with the usnic provider?

bturrubiates commented:
:+1: after some fixes to the usNIC provider.

shefty commented:
@patrickmacarthur Please let me know when this is ready to merge.  Thanks.

bturrubiates commented:
Should this PR also add the test to `runfabtests.sh`?

patrickmacarthur commented:
I squashed the commits, fixed formatting, and added the tests to the man page and `runfabtests.sh`, so this should be ready to merge.

goodell commented:
Looks like the `runfabtests.sh` changes aren't right.  Our internal Jenkins is showing this:

> **Error Message**
>
> Fail
>
> **Stacktrace**
> 
> "fi_msg_sockets -f sockets" failed
>
> **Standard Output**
> 
> SERVER OUTPUT:
> Killed by signal 15.
>
> CLIENT OUTPUT:
> Source address (-s) is required for this test

patrickmacarthur commented:
Yeah, that's because the test requires the `-s` option on both the client and server sides (since it is testing `fi_setname()`), which I'm guessing the `runfabtests.sh` script isn't set up to provide.

I can submit a separate PR to remove the test from the script for now.

goodell commented:
> I can submit a separate PR to remove the test from the script for now.

Better would be a modification to `runfabtests.sh` to also supply the client source address...
